### ISSUE TITLE
Ida Pro xmlrpc error when using pwndbg without IDA

### ISSUE THREAD

**Issue by benediktwerner** [pwndbg/pwndbg#441]
I am running `pwndbg` on WSL and don't have IDA Pro installed, which regularly leads to errors when pwndbg tries to connect to it. I just merged a fix that prevents pretty much all errors when IDA is disabled via `set ida-enabled off` (#439) but it's still quite bad if the option is not disabled, which is the default. Also, it's not that easy to find that option, so I feel like the behavior still needs some improvements.

## Some info about the errors
It seems to be mostly these two errors:
```
Traceback (most recent call last):
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/ida.py", line 64, in init_ida_rpc_client
    _ida.here()
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1098, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1437, in __request
    verbose=self.__verbose
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1140, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1152, in single_request
    http_conn = self.send_request(host, handler, request_body, verbose)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1264, in send_request
    self.send_content(connection, request_body)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1294, in send_content
    connection.endheaders(request_body)
  File "/usr/lib/python3.4/http/client.py", line 1121, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.4/http/client.py", line 951, in _send_output
    self.send(msg)
  File "/usr/lib/python3.4/http/client.py", line 886, in send
    self.connect()
  File "/usr/lib/python3.4/http/client.py", line 863, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python3.4/socket.py", line 512, in create_connection
    raise err
  File "/usr/lib/python3.4/socket.py", line 503, in create_connection
    sock.connect(sa)
InterruptedError: [Errno 4] Unterbrechung während des Betriebssystemaufrufs
```
```
Traceback (most recent call last):
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/prompt.py", line 34, in prompt_hook
    pwndbg.events.after_reload(start=False)
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/events.py", line 159, in after_reload
    f()
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/events.py", line 123, in caller
    raise e
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/events.py", line 119, in caller
    func()
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/ida.py", line 96, in __call__
    return self.fn(*args, **kwargs)
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/ida.py", line 251, in UpdateBreakpoints
    want = set(GetBreakpoints())
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/ida.py", line 227, in GetBreakpoints
    for i in range(GetBptQty()):
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/ida.py", line 96, in __call__
    return self.fn(*args, **kwargs)
  File "/mnt/d/Uni/SS18/exploits/pwndbg/pwndbg/pwndbg/ida.py", line 233, in GetBptQty
    return _ida.GetBptQty()
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1098, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1437, in __request
    verbose=self.__verbose
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1140, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1152, in single_request
    http_conn = self.send_request(host, handler, request_body, verbose)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1264, in send_request
    self.send_content(connection, request_body)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1294, in send_content
    connection.endheaders(request_body)
  File "/usr/lib/python3.4/http/client.py", line 1121, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.4/http/client.py", line 951, in _send_output
    self.send(msg)
  File "/usr/lib/python3.4/http/client.py", line 886, in send
    self.connect()
  File "/usr/lib/python3.4/http/client.py", line 863, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python3.4/socket.py", line 512, in create_connection
    raise err
  File "/usr/lib/python3.4/socket.py", line 503, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Verbindungsaufbau abgelehnt
```
The error message in German pretty much only repeats the error name. These errors mostly occur when running the application to debug for the first time, but also when trying to interrupt it with Ctrl-C to do some debugging and it seems that it also happens when meeting breakpoints. pwndbg is also sometimes lagging quite strongly which is also caused by it trying to connect to IDA Pro but timing out after three seconds. Because this happens quite often (sometimes after each error message it seems?) that leads to some extreme lag spikes in between.

Even though it's possible to just disable IDA, I still feel that the default behavior should be a bit user-friendlier and also notify the user about the option to disable the IDA integration.

I probably will try to fix this in the next few days and create a PR, but before that I would like to hear if maybe there are some other ideas and suggestions and what you think about this.

## Possible Fixes
I think the main solution would be to hide most of the stack traces and instead only display a message that the connection to IDA failed, together with a hint on how to disable it. Maybe the stack traces could only be shown if `verbose-exception` (or whatever it's called) is on. Another option might be to write them to some kind of logfile, but I guess if the option is already there this doesn't make too much sense. The main question I have here is if there are some errors that should be shown every time and also if there already is a functionality to do the "show stack-trace only if verbose-exception is on" thingy and if so how I can use it (maybe just another place where it is used or something like that).

Another thing would probably be to reduce the timeout when trying the connection to IDA a bit and maybe also make it configurable. (What do you think is a reasonable timeout?) It might also be a good idea to limit how often pwndbg tries to reconnect to IDA. (maybe only once every 10-20 seconds or something like that).

Lastly, it would be nice to have an option to disable IDA before the start of pwndbg. At the moment you can only disable it after startup as even in the `.gdbinit` file it can only be run afterwards which leads to a three second lag on every startup. One option would be a separate config file for pwndbg. While I don't particularly like the idea of an extra file it's at least better than having to wait for three seconds. This, of course, would be less of a problem if the timeout would be reduced.

## Files
If you want to see what pwndbg is currently doing just take a look at `pwndbg/ida.py`. It tries to connect with a three second timeout and if it encounters an error it prints it.

**Comment by benediktwerner** [comment]
Fail, I somehow thought I deleted the other issue I wrote when I wanted to take another look at the problem so I opened another one, even though I actually sent it. You can just ignore this 😅
